Update SonicPi to 4.5.0

porras opened the pull request:
Note that this is not the final PR, since release 4.5.0 hasn't yet been tagged in the SonicPi repo, but that will happen in the next days. The only needed change will probably the URL of the tar file (which currently points to `dev HEAD`).

All patches and build configs applied cleanly, I only needed to update the Elixir dependencies.

I built this locally and everything worked fine, but it'd be nice to see that it builds in the server.

porras committed:
Update SonicPi to 4.5.0

hfiguiere commented:
maybe it's also a good time to upgrade the runtime to `5.15-23.08`

porras commented:
@hfiguiere I just tried that locally and the build fails in a step down the road (building jack2), on first glance it could be a problem with the python version. If I have some free time on the next days I'll give it a try (maybe upgrading jack2 too helps), but that might be a rabbithole of upgrading everything :grimacing: Which makes sense and has to be done, but maybe it could be done separately. How bad would it be to release it still depending on the old runtime?

For completeness, this is the error:

```
========================================================================
Building module jack2 in /var/home/sergio/Code/net.sonic_pi.SonicPi/.flatpak-builder/build/jack2-2
========================================================================
Running: ./waf configure --prefix=$FLATPAK_DEST --htmldir=$FLATPAK_DEST/share/doc/jack/ --classic
Waf: The wscript in '/run/build/jack2' is unreadable
Traceback (most recent call last):
  File "/run/build/jack2/waflib/Scripting.py", line 141, in waf_entry_point
    set_main_module(wscript)
  File "/run/build/jack2/waflib/Scripting.py", line 191, in set_main_module
    Context.g_module = Context.load_module(file_path)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/run/build/jack2/waflib/Context.py", line 665, in load_module
    code = Utils.readf(path, m='rU', encoding=encoding)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/run/build/jack2/waflib/Utils.py", line 231, in readf
    with open(fname, m) as f:
         ^^^^^^^^^^^^^^
ValueError: invalid mode: 'rUb'
Error: module jack2: Child process exited with code 2
```

hfiguiere commented:
the `U` mode in Python has been deprecated for over a decade.

Also I just wish we'd not have to build jack at all. SonicPi is a special snowflake (to not say plain wrong)

porras committed:
Update runtime and base app

porras commented:
> Also I just wish we'd not have to build jack at all. SonicPi is a special snowflake (to not say plain wrong)

It definitely has a weird history and mix of technologies. I guess it can be forgiven by how cool it is to run it (if not to build it) :smile: 

So, as expected there were more issues with different dependencies but I managed to have something that builds locally... except it doesn't run :roll_eyes: I think this might be something specific about flatpak and wayland you might be more familiar with... maybe some change in the flatpak config is required, sockets, etc. This is the error:

```
> flatpak run net.sonic_pi.SonicPi
Starting Sonic Pi...
Failed to create wl_display (No such file or directory)
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
```

If this can't be fixed, I guess we could just merge the first commit (update Sonic Pi to 4.5.0 keeping the same runtime), which was straightforward and works, and try to fix this before the next SP release.

porras committed:
Temporarily get SP from git repo

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -364,9 +375,10 @@
+                    "//": "TODO: Point to 4.5.0 tarball when it's released",
+                    "type": "git",
+                    "url": "https://github.com/sonic-pi-net/sonic-pi.git",
+                    "branch": "dev"
```
point to a commit instead. bare branches are not allowed. even for testing

porras replied:
I just found out :roll_eyes: Makes sense but didn't think about it, since it worked locally :+1:

hfiguiere replied:
```
        "module-sonicpi-source-git-no-commit-or-tag",
        "module-sonicpi-source-git-branch"
```

hfiguiere commented:
```
Failed to create wl_display (No such file or directory)
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
```

Wayland is disabled here because it's too buggy. I don't think that's the problem, Qt QPA should figure it out, *unless this is forced upon*.

porras commented:
> Wayland is disabled here because it's too buggy. I don't think that's the problem, Qt QPA should figure it out, unless this is forced upon.

It is not forced from Sonic Pi side (the same code worked fine with the previous runtime). I guess something changed in the runtime or base app, in the way it manages that, but I have no idea how to have a look at that.

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -123,8 +123,8 @@
-                    "url": "https://github.com/jackaudio/jack2/archive/v1.9.17.tar.gz",
-                    "sha256": "38f674bbc57852a8eb3d9faa1f96a0912d26f7d5df14c11005ad499c8ae352f2"
+                    "url": "https://github.com/jackaudio/jack2/archive/v1.9.22.tar.gz",
+                    "sha256": "1e42b9fc4ad7db7befd414d45ab2f8a159c0b30fcd6eee452be662298766a849"
```
`jack_wait` that is relied upon has been removed in https://github.com/jackaudio/jack2/commit/564c710eefc760a5683b88b3ecb8a29100d5274e#diff-01eb6db751ea65e62bc79606f6cb8c728110fd84a6319ce227f29222c8c99ac0

This is why it doesn't start.

porras replied:
Oh! That message about wayland was distracting me but has been also shown when it worked, it had nothing to do with that 🙄 Cool, thanks!

hfiguiere commented:
> the `U` mode in Python has been deprecated for over a decade.

This is the same as the patch in aubio that's here.
Also

https://github.com/jackaudio/jack2/commit/328c58967dce8be426f9208ba7717ab5afc2c4f3.patch

porras commented:
Oh! That message about wayland was distracting me but has been also shown when it worked, it had nothing to do with that :roll_eyes: Cool, thanks!

porras committed:
Keep former jack2 and patch waf instead

Updating it to the latest (1.9.22) fixes the waf problem, but that version
removes `jack_wait` which Sonic Pi depends on.

hfiguiere commented:
it runs.

some weird error about 

ERROR SuperCollider:in_1 not a valid port
ERROR SuperCollider:in_2 not a valid port

but it definitely could play my stuff.

porras commented:
> it runs.

Same for me. I'll make it point to the actual 4.5.0 tarball when the release is tagged and I think it'll be good to go.

> 
> some weird error about
> 
> ERROR SuperCollider:in_1 not a valid port ERROR SuperCollider:in_2 not a valid port
> 
> but it definitely could play my stuff.

I don't know what that error means, but the current flatpak for 4.4.0 also shows it (and works).

I read they're working on improving the way all this jack stuff is done and using pipewire directly (I think they've already done for the raspberry pi version) so maybe in the last release we can get rid of jack and all this weirdness.

hfiguiere commented:
> and using pipewire directly (I think they've already done for the raspberry pi version)

If this work on Raspberry Pi then there is no reason to not enable it here. After all it's the same platform.

porras committed:
Use pipewire directly

This way we can drop the jack dependency

porras commented:
> > and using pipewire directly (I think they've already done for the raspberry pi version)
> 
> If this work on Raspberry Pi then there is no reason to not enable it here. After all it's the same platform.

Indeed. [The way they have done it](https://github.com/sonic-pi-net/sonic-pi/commit/0ef909cdba59c4f378a341b2f1dcb4e777d93d7f) has a couple of raspberry pi specific things hardcoded (the hdmi output mainly) and didn't work directly, but I wrote a small patch to do exactly what they are doing and it worked. I am not sure about whether my hardcoding of `alsa.` is correct or might break in some host systems. Since you're far more familiar with flatpak, please have a look at it. But this that I just pushed works on my machine.

In the near future I'll try to contribute this patch upstream, although I guess I'll need to rework it a bit in order to avoid break it in non-flatpak, non-pipewire linux (or convince them that that is ok). In the meantime I think it's ok to have this patch in the flatpak, isn't it?

hfiguiere commented:
my question is why the hell do they treat RaspberryPi like a special snowflake? It's basically a plain Debian (by default) or whatever other distro is run on it. It's just counter productive.

hfiguiere commented:
This works fine here.

hfiguiere commented on net.sonic_pi.SonicPi.json:
```diff
@@ -364,14 +350,19 @@
+                    "url": "https://github.com/sonic-pi-net/sonic-pi.git",
+                    "commit": "41a6728e86fb88d957690f05235d9acdfbb11aa0"
                 },
                 {
                     "type": "patch",
                     "path": "patches/sonic-pi-build-fixes.patch"
                 },
+                {
+                    "type": "patch",
+                    "path": "patches/sonic-pi-daemon.patch"
+                },
```
```suggestion
                {
                    "type": "patch",
                    "paths": [ 
                        "patches/sonic-pi-build-fixes.patch",
                        "patches/sonic-pi-daemon.patch"
                    ]
                },
```

hfiguiere commented on patches/sonic-pi-daemon.patch:
```diff
@@ -0,0 +1,82 @@
+-            Util.log "Jackd already running. Not starting another server..."
+-          end
+-        end
++        # do nothing, no need to start jack server
```
Note that even without the direct pipewire support, this would be fine, and actually might be source of the previous problem.

porras committed:
Update net.sonic_pi.SonicPi.json

Co-authored-by: Hubert Figuière <hub@figuiere.net>

porras committed:
Point to 4.5.0 release

Remove the pipewire patch, which has been merged upstream.

porras commented:
I pointed the path to the actual release. I also removed the patch, since I sent it upstream and has been accepted and merged before the release. This build works fine in my machine. If the build passes, I think it's ready to merge (probably squashing all the back and forth). @hfiguiere maybe you want to give it a final look and test.

hfiguiere commented:
thanks !
